### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    errors.push({ field: 'title', message: 'Title is required' });
+  } else if (typeof title !== 'string') {
+    errors.push({ field: 'title', message: 'Title must be a string', value: title });
+  } else if (title.trim().length === 0) {
+    errors.push({ field: 'title', message: 'Title cannot be empty' });
+  } else if (title.length > 100) {
+    errors.push({ field: 'title', message: 'Title must be 100 characters or less' });
+  } else if (/<[^>]*>/g.test(title)) {
+    errors.push({ field: 'title', message: 'Title cannot contain HTML tags' });
+  }
```
Comment (dandeliongold) [inline src/middleware/validation.ts:34]: In `src/middleware/validation.ts` the description regex uses `/<.*>/g` while title uses `/<[^>]*>/g`. The `/<.*>/` pattern is greedy and can behave differently (it will match across tags and may allow some tags). Use a single consistent, safe regex (e.g. `/[^<>]+/` checks or better, use a proper sanitizer library) to avoid inconsistencies and false negatives/positives.

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<[^>]*>/g.test(description)) {
+      errors.push({ field: 'description', message: 'Description cannot contain HTML tags' });
+    }
+  }
+
+  // Validate completed if provided
```
Comment (dandeliongold) [inline src/middleware/validation.ts:85]: The `validateTodoId` allows `0` because the check only ensures digits with `^\d+$`. If your IDs are 1-based, explicitly reject `0` (e.g. `!/^[1-9]\d*$/.test(id)`) and add a test for the `"0"` edge case.

File: src/routes/todos.ts [hunk 3]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
   const { id } = req.params;
   const todo = todoStore.getById(id);
   
   if (!todo) {
-    return res.status(404).json({ error: 'Todo not found' });
+    return res.status(404).json({ 
+      error: 'Todo not found',
+      details: [{ field: 'id', message: `Todo with ID ${id} does not exist`, value: id }]
+    });
   }
```
Comment (dandeliongold) [inline src/routes/todos.ts:23]: In `src/routes/todos.ts` the `post` handler calls `todoStore.create()` without try/catch. If `todoStore.create` can throw (IO, assertion, etc.) the process could crash. Wrap the store calls in try/catch and return a 500 JSON error with a safe message, or ensure `todoStore` never throws.

File: package.json
Comment (dandeliongold) [inline package.json:1]: The PR adds `supertest` and `@types/supertest` to `package.json` and `package-lock.json`. These are fine, but please run `npm install` and ensure `npm test` passes on CI. Consider shrinking package-lock changes by only adding the minimal dev deps rather than many transitive additions.

**Review by dandeliongold — CHANGES_REQUESTED** [review]
Thanks for the thorough validation feature — this improves safety and test coverage. I have a few requested changes before merging:

- Make HTML-tag detection consistent (fix the description regex or use a sanitizer).
- Reject ID "0" explicitly if IDs are 1-based and add tests for it.
- Add try/catch around store operations or ensure store methods never throw; return safe 500 responses on unexpected errors.
- Add the missing tests noted in `validation.test.ts` (description HTML test and ID "0").
- Re-run `npm install` and CI to confirm tests pass; consider trimming lockfile churn if possible.

I left inline comments with details and suggestions. Once addressed I’ll re-review and approve.
